### ISSUE TITLE
New Capsul API

### ISSUE THREAD

**Issue by sapetnioc** [populse/capsul#198]
There are some major changes that are ongoing in Controller. With the need to get rid of `StudyConfig` class, this will imply a major change in Capsul API. This issue is here to centralize all the commits and discussions related to this API changes.

**Comment by denisri** [comment]
I was just about to start a discussion on soma-base...
Will a single ticket be enough for all discussions ?
I am starting to look at the new `Controller`, and I have difficulties to use them, but probably because I'm not familiar with the typing syntax of python3 (?), and also lack of doc in `Controller` right now. Should we discuss all here, or the Controller part in [soma-base](https://github.com/populse/soma-base) ?

**Comment by servoz** [comment]
When do you plan to defenitely get rid of StudyConfig?

The current version of the controller can only handle simple types, but, as already said in mia's tickets, some bricks may require a controller handling complex types (dict of dict of list of list of tuple, that's the first idea that comes to my mind and I'm not sure it's a real case we need, it's just to illustrate...).
We made a patch in mia to go back to version 1 of the controller (which can potentially handle all complex types) if needed.
I'm not sure that the new controller will be able to handle all possible types anyway, from what I could see when I read this part of the code, but it's been a while now and with the new version coded by @sapetnioc  maybe it's different ?
Not to bore you with this and because we have the patch back to version 1, I haven't discussed too much about the controller that I think would be needed... But I can to do it when it will be the good time !

**Comment by sapetnioc** [comment]
@denisri I did this ticket for commits of the `pydantic_controller` branch. I think `Controller` related discussion can go in soma-base.

**Comment by denisri** [comment]
> When do you plan to defenitely get rid of StudyConfig?
> 
> The current version of the controller can only handle simple types, but, as already said in mia's tickets, some bricks may require a controller handling complex types (dict of dict of list of list of tuple, that's the first idea that comes to my mind and I'm not sure it's a real case we need, it's just to illustrate...). We made a patch in mia to go back to version 1 of the controller (which can potentially handle all complex types) if needed. I'm not sure that the new controller will be able to handle all possible types anyway, from what I could see when I read this part of the code, but it's been a while now and with the new version coded by @sapetnioc maybe it's different ? Not to bore you with this and because we have the patch back to version 1, I haven't discussed too much about the controller that I think would be needed... But I can to do it when it will be the good time !

You are speaking about the GUI I guess ? `ControllerWidget`, not `Controller` itself ?
In this ticket discussion will deal with the branch `pydantic_controller`, which involves a new GUI, yes, but most of all a drastic change in API: remove traits and use python3 typing syntax instead. @sapetnioc has been working on it, and I am just starting to look at it. All the `Process`/`Pipeline` API will change, and nipype bridges will need to be re-written.

**Comment by denisri** [comment]
> @denisri I did this ticket for commits of the `pydantic_controller` branch. I think `Controller` related discussion can go in soma-base.

OK.

**Comment by sapetnioc** [comment]
@servoz we will do all the major changes at the same time. This is in the  `pydantic_controller` branch of both `soma-base` and `capsul`. As Denis mentioned there is a lot of work to do. It is still difficult to say when it will be ready but it will be at least in months 

the new controllers supports already types constructed with `list`, `dict`, `Union` and `Literal` (new in Python 3.9). For instance you can use `dict[Literal('one', 'two', 'three'), Union[str, int, float]]`. Not all combinations will be supported by GUI (espacially for dict keys) but it should cover all our needs.

**Comment by servoz** [comment]
yes I'am speaking about the ControllerWidget ... and this is a class of soma ... Sorry I wrote too fast ...

**Comment by sapetnioc** [comment]
The widgets are not finished but there is already a better support for recursive types (e.g. list[list[list[Controller]]]). I still have to work on Union and the possibility to add elements to dictionaries. But the code is simpler it is a matter of one week or two (but full time which never happens).

**Comment by servoz** [comment]
> The widgets are not finished but there is already a better support for recursive types (e.g. list[list[list[Controller]]]). I still have to work on Union and the possibility to add elements to dictionaries. But the code is simpler it is a matter of one week or two (but full time which never happens).

OK

**Comment by denisri** [comment]
I copy a discussion started on https://github.com/brainvisa/use-cases/issues/1, which is better placed here:

@denisri
So there is a new object named `Capsul` in the API. It apparently replaces the `CapsulEngine.get_process_instance()` method with `Capsul.executable()` and contains the `CapsulEngine` object we had before. Was it useful to separate it into 2 classes, which, with `Settings`, makes 3 "global" objects with a public API ? Doesn't it make it a bit difficult for a novice user (programmer) to start, where a single object would have gathered the main user API in a single place ?

@sapetnioc
This is to be discussed. I made these choices without a written use case so it is not obvious that this is optimal. I put some explanation in the comments. The Capsul object is similar to application objects in web frameworks or Qt. Its main goal is to read all the configurations accessible to user : site level (e.g. on /etc/...), user level (e.g. on /home/...) and session level (e.g. environment variables). In this configurations, user may have access to several capsul engines, for instance local, triscotte and n4h. The separation allows to easily choose one of them for processings:
```
capsul = Capsul()
with capsul.engine('triscotte') as capsul_engine:
    # Do something on triscotte here via capsul_engine
```
I do not think we need a `Settings` objects here (configuration could be managed by `Capsul` API) but there are 3 still different concepts in this code:

    `Capsul` that exists only in user environment
    `CapsulEngine` that makes the link between user environment and execution environment
    `Executable` that exists in both environments. To date, it is just another name for Process instance to be able to clearly identify when we talk about a process (i.e. a kind of piece of code that needs parameters to run) from the executable that contains a process as well as all it needs to be executed (e.g. its parameter values).

@denisri
OK we'd better have this discussion in the capsul project. Let's move to populse/capsul#198.

@sapetnioc
In my example, I tend to use enough details to be able to discuss internal details. But I keep in mind that we will probably provide some useful shortcut methods. For instance, the detailed example may easily be made compatible with the following code:
```
from capsul.api import capsul_engine

with capsul_engine('triscotte') as ce:
    executable = ce.run('bv_use_cases.simplest.list_directory')
    print(executable.result)
```

**Comment by denisri** [comment]
So the objects seem to be somewhat shifted compared with the capsul/master objects:
`Capsul` is more or less what `CapsulEngine` is in master
`CapsulEngine` replaces (more or less, and for now, includes) what `Settings` is in master
with some notable differences:
We used to have a single `Settings` object, which was expected to store all configurations. Then we could "activate" one of them and process things (on computing server side, or "executing environment"). On client side we were doing most of the job regardless of which config will be selected, or at least, as long as possible: we could instantiate pipelines, perform parameters completion, etc.
Here the config activation seems more central, and we have (or can create) a separate `CapsulEngine` object for each configuration. Why not, it's just different.

I just wonder about the `with` statement: in my mind (I can be wrong) it's better to use `with` for limited portions of code, with a limited execution time, because it's usually used to "lock" an object. In master, it as used just the time needed to modify a `Settings` configuration, and for execution basically we just have to specify which configured resource we want to use. I mean, we can disconnect the client while running, and reconnect later, thus the whole execution is not forced to be locked within a `with`... statement, does it ? So: when exactly is `with` needed ?

**Comment by denisri** [comment]
Another point, or maybe discussion, is more general:
in the `pydantic_controller` branch, we thus don't have "just" the traits API replaced with python3 syntax and pydantic. There is also the removal of the old `StudyConfig`, but also a very deep refactoring of the "capsul v3" we were previously working on. I don't say it's bad, it's a real question for myself as well, but I wonder if we really can afford such huge changes, which will likely involve a mostly complete rewriting of Capsul. There has already been years of efforts to get where we are now, and it's not even pushed in production yet before we start it all over again. I mean, we can't wait years again before having something usable, and working in the new branch will absorb most of our efforts, thus we will not push the current master branch into production state as it is already obsolete. I understand the will to improve things and propose an "always better" architecture, but at some point we must deliver something usable on which developers can rely and users can use. Axon is more or less abandoned and unmaintained, capsul/master works but lacks a few features so is not in production yet, and we're about to drop it also, with nothing ready to replace the pipelining infrastructure. Thus I fear we "never" reach something stable enough, and that users/developers just give up. They are expecting something working **soon** (and so do I), so maybe we'd better focus on more limited changes (at first) that we can finish within a few months, not more ?
As I have said, it's a question, my opinion is not very clear...

**Comment by sapetnioc** [comment]
Due to BrainVISA backend discussions, some decision taken for the master API seemed to be wrong. An important one is to rely only on PopulseDB to store configuration. One of the strong requirements was to be able to use Capsul() without any databasing system, even for parameter completion. To my opinion this will force us to rethink the configuration API and corresponding backend. This is why settings have disappeared for now. To be honest, I had no enough time to have a clear view of what we can keep and what we should drop from master branch on this topic.

Another thing we decided is to have a clear separation between the user environment and the execution environment. Processes are never executed in the same Python as the one launched by the user. Hence, there is no more need to be able to activate a configuration in the current Python. This will have an influence on the API.

Regarding the "with" usage. I wanted it to be only necessary when a connection to the execution environment is required (for using `start()`, `wait()`, etc.). I'd like parameter completion to be able to work outside the with. But since no specific engine is selected before the with and it may be necessary to have it for the completion, I am probably missing something in the API.

I think we need to write a new use case to precise what features we need for configuration. It could be something, like : "Configuration allows me to use 3 execution engines. One defined and maintained by my lab system administrator. One cluster accessible through ssh with my account and my local computer.  I want to be able to configure how I use these engines (e.g. limit CPU number) and to save these configuration at my account level. I must be able to give a label to each configuration and to use a fully configured engine in Capsul by just giving one label."

**Comment by sapetnioc** [comment]
I agree with your general comment. There are important pros and cons to these changes. You pointed out the major cons : huge changes. But for the pros, I would add the following:

- We must have a more secure solution to go through internet. This is a need for some of our platforms.
- We cannot maintain existing infrastructure for more than a few years. For instance, I think there are only one or two people that can setup soma-workflow
- We may have more people participating if we explain from the begining and clean things up

But I share your fears. If we do not do something we believe in quickly, it will mean that this path was not the good one.

**Comment by denisri** [comment]
Well, I agree in large parts, but still fear the amount of work and still try to reduce the changes, or at least, get there by steps ;) The thing is that our work force is very limited: concretely you and me are the only people I am sure will spend time on it (others are welcome of course, but who will actually ?), and neither you or me can work on it full-time, so we'd better fraction the work in small milestones, and, if possible, not start all the works at the same time.

> Due to BrainVISA backend discussions, some decision taken for the master API seemed to be wrong. An important one is to rely only on PopulseDB to store configuration. One of the strong requirements was to be able to use Capsul() without any databasing system, even for parameter completion. To my opinion this will force us to rethink the configuration API and corresponding backend. This is why settings have disappeared for now. To be honest, I had no enough time to have a clear view of what we can keep and what we should drop from master branch on this topic.

I agree for populse_db, but if I'm right it's only used to store settings, possibly internally (in memory), and mostly under the hood (so can be reimplemented transparently). Completion doesn't use databasing at all so doesn't require an immediate API refactoring, even if it would be good to simplify the completion API, but it's an uncorrelated matter.
Config values are serialized in json and passed to the computing resource in json for execution, so the database is maybe superfluous, but not an API dead-end in the master branch implementation.

> Another thing we decided is to have a clear separation between the user environment and the execution environment. Processes are never executed in the same Python as the one launched by the user. Hence, there is no more need to be able to activate a configuration in the current Python. This will have an influence on the API.

That's partly what has motivated my question about the `with` statement: in this context I'm not sure it is really useful on client side ?

> Regarding the "with" usage. I wanted it to be only necessary when a connection to the execution environment is required (for using `start()`, `wait()`, etc.). I'd like parameter completion to be able to work outside the with. But since no specific engine is selected before the with and it may be necessary to have it for the completion, I am probably missing something in the API.

I totally agree.
`with` looks nice and elegant in small code snippets, but it has strong limitations: it's a code block, so you can't get out of scope without releasing/disconnecting/deleting the object you are using. Soma-workflow client for instance allows to have several simultaneous connections (to several computing resources), which is impossible (or at least difficult) to do in a `with` block. That's just an example.
I think it's good to lock configuration settings while they are being modified in a multi-threaded context (with or without a database), for instance, but I'm not so sure for computing resources connections.

> I think we need to write a new use case to precise what features we need for configuration. It could be something, like : "Configuration allows me to use 3 execution engines. One defined and maintained by my lab system administrator. One cluster accessible through ssh with my account and my local computer. I want to be able to configure how I use these engines (e.g. limit CPU number) and to save these configuration at my account level. I must be able to give a label to each configuration and to use a fully configured engine in Capsul by just giving one label."

Yes...

**Comment by sapetnioc** [comment]
Well, there are many things that are work in progress. I used to be alone on this branch and I thought it would be shared later (in a more stable state).

The definition is the string to use to serialize a process. This is meant to be passed to `executable()`. I also started some changes about this string that I wanted to discuss here. For your question, it is not always possible to find this definition string from the executable content. For instance if the process is defined in an XML file. Therefore I choose to manage this string from the begining. It is `executable()` that pass the value to the various constructors (Process, Pipeline, etc.).

**Comment by denisri** [comment]
> Well, there are many things that are work in progress. I used to be alone on this branch and I thought it would be shared later (in a more stable state).

OK. Do you prefer us to wait for you to stabilize a few things before looking at it ?

> The definition is the string to use to serialize a process. This is meant to be passed to `executable()`. I also started some changes about this string that I wanted to discuss here. For your question, it is not always possible to find this definition string from the executable content. For instance if the process is defined in an XML file. Therefore I choose to manage this string from the begining. It is `executable()` that pass the value to the various constructors (Process, Pipeline, etc.).

OK but then let's make it optional, and let the class guess the definition by default: in a regular .py module file it should be able to set it automatically.

**Comment by sapetnioc** [comment]
Regarding the process definition string, here is a summary of what it accepts to date:

- A python object located in an importable module. The definition string must have the pattern `{module}.{attribute}`. This is a subset of what we have in master. I dropped the possibility to use only the module name because any function may be a process and I do not know how to distinguish potential process function from a standard function. Therefore finding the single process in a module could only work if there is no function in the module. It became too complicated. The supported Python objects are classes and instances of `Process` and Nipype `Interface` (not implemented) as well as functions with annotations (i.e. parameters typing).
- A JSON file. I propose to drop XML format in favor of JSON. I have web services in mind where we can directly use this format. If there are any pipeline in XML, we may either make a converter or support the format for backward compatibility. In all cases, the definition string must accept a full file path. This is not a relocatable value so I will likely reintroduce a definition based on a Python module directory. If possible, I would like the definition string to be enough to identify the structure of the definition. For JSON, I have in mind that a string ending with `.json` would always be a json file. Therefore `my.module.my_process.json` would indicate the `my_process.json` file located in the directory of the `my.module` Python package.

For pipelines, the definition would be `{pipeline definition}#{node name}` (in `pipeline_definition` branch). Therefore `morphologist#split_brain#threshold` would be the `threshold` process of the `split_brain` sub-pipeline of `morphologist` pipeline. I do not know yet if `executable()` should accept these definitions. I do not see good use cases to extract a process from a pipeline. But pipeline construction gives this definition string to process nodes.

**Comment by sapetnioc** [comment]
If the only entry point to create a process is `executable()`, I think `definition` can be mandatory. This would be the job of the function to set it. This prevent the direct use of a process class but it is on purpose. Since processes are now always executed in a separate Python, its class must be serializable. But we cannot serialize any class (for instance a class defined in a function) so it have to be in a module. Forcing the use of `executable()` with the module name and class name (in definition string) is a way to be sure that the process can be serialized for its execution.

**Comment by sapetnioc** [comment]
I am being working on pipeline definition. These API changes should be more coherent at the end of this work. Now, I only have uncommented a single pipeline test (in `pipeline_definition` branch, there is an associated draft PR).

**Comment by denisri** [comment]
> Regarding the process definition string, here is a summary of what it accepts to date:
>     * A python object located in an importable module. 

**or a .py file**. We could use:
```
pipeline = engine.get_process_isntance('/tmp/test_pipeline.py#MyPipelineClass')
```
this was quite useful, why drop this possibility ?

> The definition string must have the pattern `{module}.{attribute}`. This is a subset of what we have in master. I dropped the possibility to use only the module name because any function may be a process and I do not know how to distinguish potential process function from a standard function. Therefore finding the single process in a module could only work if there is no function in the module. It became too complicated. The supported Python objects are classes and instances of `Process` and Nipype `Interface` (not implemented) as well as functions with annotations (i.e. parameters typing).

The shortcut was handy. I have been writing thousands of time `engine.get_process_instance('morphologist.capsul.morphologist')` which is already long enough, the need to add the unnecessary `.Morphologist` is not so negligible.
Could we use the single process detection for `Process` classes only ? It would not work for a function, but it's a less common and more ambiguous usecase.

>     A JSON file. I propose to drop XML format in favor of JSON. I have web services in mind where we can directly use this format. If there are any pipeline in XML, we may either make a converter or support the format for backward compatibility. In all cases, the definition string must accept a full file path. This is not a relocatable value so I will likely reintroduce a definition based on a Python module directory. If possible, I would like the definition string to be enough to identify the structure of the definition. For JSON, I have in mind that a string ending with `.json` would always be a json file. Therefore `my.module.my_process.json` would indicate the `my_process.json` file located in the directory of the `my.module` Python package.

I agree for json vs XML. The syntax, however, is a bit ambiguous: `my.module.my_process.json` could also be a `json` module or class inside in `my_process`...

> For pipelines, the definition would be `{pipeline definition}#{node name}` (in `pipeline_definition` branch). Therefore `morphologist#split_brain#threshold` would be the `threshold` process of the `split_brain` sub-pipeline of `morphologist` pipeline. I do not know yet if `executable()` should accept these definitions. I do not see good use cases to extract a process from a pipeline. But pipeline construction gives this definition string to process nodes.

Why not.

**Comment by sapetnioc** [comment]
If one wants to see the current state of file name completion with temporary file names management, use the following command: 
```
python -m bv_use_cases.tiny_morphologist
```
This is still incomplete (extensions are missing from use case) and buggy (especially for switch) but it gives an idea of the final behavior. Now parameter completion and tmp value setting is done in the `set_output_paths()` method of Dataset but it will probably change in the future.

**Comment by denisri** [comment]
In the new pipeline implementation, how and when is exactly supposed to happen parameters values propagation through links ? Right now if I set a value on an input parameter of a pipeline, it does propagate to nodes directly connected to it, but if I set a value on an inner node (for instance an output value of a node), then it does not propagate to downstream nodes. Is it normal with a "manual" propagation (how ?), is it something not finished, or is it a bug ? In the pipeline implementation I don't see any callback for propagation so I feel it is not implemented at all (but then how does the pipeline node values propagation to directly connected nodes happen ?)

**Comment by denisri** [comment]
Thoughts about values propagation:
After discussion with @sapetnioc we were on the idea to operate values propagation on demand, whereas it was done automatically on-the-fly in capsul v2. Doing it on demand would avoid useless processing because many values propagate several times through many links. But when thinking a little bit more, it appears that values propagation is not a well-defined notion, especially if it is not done upon assignation as before.
* which parameters values should propagate ? When values are not the same at both ends of a link, which side is the correct one ?
  * output -> input ? not really: some values are set on output parameters (file names) and should propagate backward to upstream inner nodes of a pipeline
  * pipeline main node -> inner nodes ? Not always: sometime we expect pipeline values to be the result of an inner link, especially when switches are involved
* switches, and several other custom nodes, are meant to propagate values in either direction, depending on which values is set externally on them. Thus the action (and the moment) of setting a value, is important, and cannot be guessed afterwards.

Maybe these problems are design flaws because things are not completely determined from a values dict for instance, and they need some dynamic information, which is neither documented, nor desirable. This can cause problems when saving/reloading pipelines parameters values, for instance.
However I don't see an obvious solution for them right now, and it "happened to work reasonably in most situations" in v2, which also suggests that such ambiguous situations were not so frequent, but there is something missing anyhow.

**Comment by denisri** [comment]
I am reading the `Dataset` code for path completion. It's a start, why not, but it needs to be completed and extended: right now completion for a pipeline or process is operated from a single dataset, which fills all of its outputs, and only outputs.
In "real life" we have processes taking several inputs from the same dataset (for instance a brain mesh process in Morphologist will use, for instance, a bias corrected T1 image, an hemisphere segmentation, maybe a histogram analysis... So we need completion on inputs also.
Moreover we also have cross-datasets pipelines, which will use data from a Morphologist segmentation, a Freesurfer segmentation, and a tractography (FSL, Connectomist, MRTrix...) which need to be combined to deduce output path names.
I don't think we have real life examples of several output datasets for the same pipeline, I don't know if this will be useful.
Anyway I think we need a way to indicate, for a given process, several datasets associated with several of its parameters. How can we do that ?

Otherwise I'm not sure to be totally convinced (yet) by the way some layout attributes are set by processes: I'm thinking of the "side" attribute we used in Morphologist, which is now handled as a "prefix" in the tiny_morphologist example. I'm not familiar with the system yet so maybe it's OK, but I feel using "prefix" is a kind of "bidouille" which may indicate that something is missing in the paths definition: I think "side" deserves a real conceptual existence, since it is used in many "lateralized" processings. But it's possibly a matter of naming attributes...

**Comment by denisri** [comment]
To overcome the multiple Dataset / layout issue, I have implemented a proposition, in a class temporarily called `Completion`. It allows to assign any `Dataset` instance to processes/parameters. It thus needs to be configured before use, at the moment. Otherwise it works much like `Dataset.set_output_paths()` but using possibly different datasets for different processes/parameters, and it does not filter out input parameters. It is operated using `Completion.set_paths(process, **kwargs)`.
I have updated the _tiny_morphologist_ example in `bv_use_cases` to use it - in this example the result is the same.

**Comment by denisri** [comment]
I wonder if we haven't been too ambitious by starting refactoring and rewriting everything at the same time in soma-base/Controller and Capsul - I feel a bit lost now. We have started works on all aspects at the same time:
1. moving away from traits
2. redesigning the pipeline API
3. removing StudyConfig
4. refactoring configuration and CapsulEngine
5. building a new completion system
6. refactoring the execution engine

I would have done one thing after another, but I realize all these aspects are already started, former implementations have already been removed, thus are broken, none is finished, and I don't really see a path to have things working in a near future...

For instance, the new completion system is possibly a good path to go, but it's far from being finished: it doesn't handle lists, it's not integrated with iterations and workflows etc. There has been a significant amount of work in the previous implementations, and it was perhaps a bit naive to think we would redo all of it within a few weeks... ?
I think we should plug a number of old things back in the meantime, typically the older completion engine and the execution engine could be ported to the new APIs and rewritten later ? I don't really know about the state of the config system now, but all former config modules are gone. Maybe we should focus on points 1-3 first ?

**Comment by patafana** [comment]
I've noted that, in the new API, the PipelineNode class inside [pipeline_nodes.py](https://github.com/populse/capsul/blob/master/capsul/pipeline/pipeline_nodes.py) isn't present anymore. Is there a replacement object with equivalent features?

**Comment by denisri** [comment]
Indeed, in capsul v2, `Node` and `Process` were different objects. In a pipeline, a `Process` was inserted in a `ProcessNode`, a `Pipeline` was inserted in a `PipelineNode`. In v3 we have chosen to make `Process` directly inherit `Node`, thus we cannot dissociate the `Node` part and the `Process` part, but it simplifies many things in the codes. Thus `PipelineNode` is not needed any longer, and the "equivalent" is just `Pipeline`.

**Comment by sapetnioc** [comment]
I am closing this issue because the 3.0 API design is finished. Future changes could be followed via more specific issues.
